btc-lsp/src/BtcLsp/Thread/Refunder.hs

```diff
@@ -0,0 +1,138 @@
+toHex = decodeUtf8 . B16.encode
+
+data SendUtxoConfig = SendUtxoConfig
+  { dustLimit :: MSat,
+    estimateFee :: MSat,
+    satPerVbyte :: Integer
+  }
+
+defSendUtxoConfig :: SendUtxoConfig
+defSendUtxoConfig =
+  SendUtxoConfig
+    { dustLimit = MSat $ 20000 * 1000,
+      estimateFee = MSat $ 500 * 1000,
+      satPerVbyte = 1
+    }
+
+debugLog :: (KatipContext m) => Text -> m ()
+debugLog = $(logTM) DebugS . logStr
+
+errorLog :: (KatipContext m) => Text -> m ()
+errorLog = $(logTM) ErrorS . logStr
+
+getOutPoint :: (OP.OutPoint, MSat, Maybe ByteString) -> OP.OutPoint
+getOutPoint (u, _, _) = u
+
+getAmt :: (OP.OutPoint, MSat, Maybe ByteString) -> MSat
+getAmt (_, a, _) = a
+
+getLockId :: (OP.OutPoint, MSat, Maybe ByteString) -> Maybe ByteString
+getLockId (_, _, l) = l
+
+sendUtxosWithMinFee ::
+  (Env m) =>
+  SendUtxoConfig ->
+  [(OP.OutPoint, MSat, Maybe ByteString)] ->
+  Text ->
+  Text ->
+  ExceptT Failure m (Btc.DecodedRawTransaction, MSat, MSat)
+sendUtxosWithMinFee cfg utxos addr txLabel = do
+  when (estimateAmt <= dustLimit cfg) $ liftIO $ fail "Total utxos amount is below dust limit"
+  mapM_
+    ( \r -> case getLockId r of
+        Just l -> void $ withLndT releaseOutput ($ RO.ReleaseOutputRequest l (Just $ getOutPoint r))
+        _ -> pure ()
+    )
+    utxos
+  ePsbt <- withLndT fundPsbt ($ ePsbtReq)
+  finPsbt <- withLndT finalizePsbt ($ FNP.FinalizePsbtRequest (FP.fundedPsbt ePsbt) "")
+  decodedETx <- withBtcT Btc.decodeRawTransaction ($ toHex $ FNP.rawFinalTx finPsbt)
+  let fee = MSat $ fromInteger (satPerVbyte cfg * 1000 * Btc.decVsize decodedETx)
+  void $ releaseUtxosPsbtLocks (FP.lockedUtxos ePsbt)
+  let rPsbtReq = fundPsbtReq utxos addr fee
+  rPsbt <- withLndT fundPsbt ($ rPsbtReq)
+  finRPsbt <- withLndT finalizePsbt ($ FNP.FinalizePsbtRequest (FP.fundedPsbt rPsbt) "")
+  decodedFTx <- withBtcT Btc.decodeRawTransaction ($ toHex $ FNP.rawFinalTx finRPsbt)
+  ptRes <- withLndT publishTransaction ($ PT.PublishTransactionRequest (FNP.rawFinalTx finRPsbt) txLabel)
+  if null $ PT.publishError ptRes
+    then pure (decodedFTx, totalUtxoAmt, fee)
+    else liftIO $ fail "Failed to publish refund transaction"
+  where
+    ePsbtReq = fundPsbtReq utxos addr estimateAmt
+    estimateAmt = totalUtxoAmt - estimateFee cfg
+    totalUtxoAmt = sum $ getAmt <$> utxos
+    releaseUtxosPsbtLocks lutxos =
+      mapM_ (\r -> withLndT releaseOutput ($ toROR r)) lutxos
+      where
+        toROR (FP.UtxoLease id' op _) = RO.ReleaseOutputRequest id' (Just op)
+    fundPsbtReq utxos' outAddr fee = do
+      let amt' :: Word64 = coerce (totalUtxoAmt - fee)
+      let r = MSat $ (* 1000) $ fromInteger $ ceiling (toRational amt' / 1000)
+      let mtpl = FP.TxTemplate (getOutPoint <$> utxos') (M.fromList [(outAddr, r)])
+      FP.FundPsbtRequest "" mtpl 2 False (FP.SatPerVbyte 1)
+
+sendUtxos ::
+  (Env m) => [(OP.OutPoint, MSat, Maybe ByteString)] -> Text -> Text -> ExceptT Failure m (Btc.DecodedRawTransaction, MSat, MSat)
```

Comment:
Two blind Text and MSat parameters, fix with newtypes and data/records instead of 3 element tuples